// graphql-schema-linter/lib/find_schema_nodes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findSchemaNodes = findSchemaNodes;

var _graphql = require("graphql");

/*
  scopes: (required) string array
  schema: (required) GraphQLSchema

  A scope could be a `Type`, `Type.field`, `Type.field.argument` or `Enum.VALUE`.
*/
function findSchemaNodes(scopes, schema) {
  const result = new Set();
  const tracer = {
    enter: node => {
      result.add(node);
    },
    leave: () => {}
  };

  for (const scope of scopes) {
    const node = findScopeNode(scope, schema);
    node && (0, _graphql.visit)(node, tracer);
  }

  return result;
}

function findScopeNode(scope, schema) {
  var _astNode;

  const [typeName, fieldName, argumentName] = scope.split('.');
  const type = schema.getType(typeName);
  let astNode = type === null || type === void 0 ? void 0 : type.astNode;

  if (fieldName === undefined) {
    return astNode;
  }

  const field = ((_astNode = astNode) === null || _astNode === void 0 ? void 0 : _astNode.kind) === 'EnumTypeDefinition' ? type === null || type === void 0 ? void 0 : type.getValue(fieldName) : type === null || type === void 0 ? void 0 : type.getFields()[fieldName];
  astNode = field === null || field === void 0 ? void 0 : field.astNode;

  if (argumentName === undefined) {
    return astNode;
  }

  const argument = field === null || field === void 0 ? void 0 : field.args.find(arg => arg.name === argumentName);
  astNode = argument === null || argument === void 0 ? void 0 : argument.astNode;
  return astNode;
}
